### PR TITLE
[Issue #1337] E2E Testing ADR

### LINKED ISSUE #1337: [Task]: Look into E23 / Integration Testing
### Summary

There are a couple of choices with an E2E / integration testing framework.

This task is to review a couple of top choices, determined by the implementer. The work from this investigation could be in the form of a comment. This could be moved to an [ADR](https://github.com/HHS/simpler-grants-gov/tree/main/documentation/decisions#architecture-decision-records-adrs). Bonus points for orienting the research toward creating an E2E ADR.

### Acceptance criteria

- [ ] Identify best open source E2E / integration options
- [ ] Provide some basic pros and cons

### PR DESCRIPTION
**Pull request by kniedermaier**
## Summary
Fixes #1337

## Changes proposed
Adds an ADR to use Playwright for end-to-end / integration testing

## Context for reviewers
Compares Cypress and Playwright

### CODE REVIEW

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+Playwright's modern syntax, variety of tracing tools, and speed makes for an ideal developer experience, which increases in the case that a developer's IDE of choice is VS Code. It also supports a variety of browsers and platforms and efficiently spins up new browser contexts for each test. Cypress has considerable documentation and an extensive commmunity, which more than makes up for how it handles async/sync code. If there isn't a specific need that warrants E2E tests be run in browser or adversion to a Microsoft-backed open-source project, then the performance gains that Playwright offers is reason enough to choose it as the E2E framework.
+
+## Pros and Cons of the Options
+
+### Cypress
+
+#### Pros
+- Runs in browser, which means test code is evaluated with JavaScript as opposed to Node or another server-side language and there's native access to whatever portion of the application is being tested
+- Extensive support for reusability, including creating shortcuts to recreating specific application states
+- Plenty of examples and tutorials to refer to as this framework has been in existence for many years
```
Comment (rylew1) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:36]: This is pretty important imo

Comment (rylew1) [reply documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:36]: What's the debugging like for cypress?

Comment (kniedermaier) [reply documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:36]: Cypress has a desktop app (https://docs.cypress.io/guides/core-concepts/cypress-app) where you can use  browser dev tools like debugger or step through the test. Some of the desktop app functionality relies on having a Cypress Cloud (enterprise) account in place.

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+- Runs in browser, which means interacting with a DB or other backend service requires extra work. For instance, methods would need to be exposed in order to seed a DB.
+- Cypress handles chains of commands such that mixing async and sync code is difficult to read and understand without familiarity of its syntax (e.g. promise chaining as opposed to a more readable async/await approach)
+- Limited iFrame, multiple tabs, and hover support
+- Testing on Safari browsers is experimental and relies on Playwright
+
+### Playwright
+
+#### Pros
+- Tests are run in parallel by default and benchmark speeds outpace Cypress
+- Can run tests in headless, headed, or UI mode depending on use case and on Windows, Linux, and macOS with support for Chromium, WebKit and Firefox
```
Comment (widal001) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:48]: This flexibility seems like a huge pro!

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+- Supports native mobile emulation of Google Chrome for Android and Mobile Safari
+- Backed by Microsoft and has a powerful VS Code extension to debug tests within the IDE
```
Comment (widal001) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:50]: VSCode extension also seems like an important differentiator

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md (hunk 2 above)
Comment (widal001) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:48]: This flexibility seems like a huge advantage!

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+- Uses async/await syntax and locator methods return elements
+
+#### Cons
+- Many of the debugging and ease-of-use functionality leverages VS Code, which means developers who prefer another IDE might be less effective at debugging tests
```
Comment (widal001) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:54]: Do we have a sense for the current split on the team of IDEs that they use?

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# E2E / Integration Testing Framework
+
+- **Status:** {Accepted}
```
Comment (widal001) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:3]:
```suggestion
- **Status:** Accepted
```
Just removing the extra brackets around "Accepted"

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+- **Last Modified:** 2024-02-26
+- **Related Issue:** [#1337](https://github.com/HHS/simpler-grants-gov/issues/1337)
+- **Deciders:** {other collaborators}
```
Comment (widal001) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:6]: Can we add the folks who weighed in on this decision before we merge it in?

**Review by widal001 — APPROVED** [review]
LGTM! Left a few minor comments and suggestions, but otherwise this seems like a sensible choice.

Thanks for documenting this comparison and recommendation.

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md (hunk 4 above)
Comment (kniedermaier) [reply documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:54]: I'm not sure, and I'm happy to poll the team via slack.

File: documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md (hunk 6 above)
Comment (kniedermaier) [inline documentation/decisions/adr/2024-02-26-e2e-integration-testing-framework.md:6]:
```suggestion
- **Deciders:** Billy Daly, Ryan Lewis
```

**Pull request merged**